internal/action/diff.go

```diff
@@ -119,5 +124,125 @@ func Diff(ctx context.Context, config *helmaction.Configuration, rls *helmreleas
 	if err != nil {
 		errs = append(errs, err)
 	}
-	return set, errors.Reduce(errors.Flatten(errors.NewAggregate(errs)))
+	return set, apierrutil.Reduce(apierrutil.Flatten(apierrutil.NewAggregate(errs)))
+}
+
+// ApplyDiff applies the changes described in the provided jsondiff.DiffSet to
+// the Kubernetes cluster.
+func ApplyDiff(ctx context.Context, config *helmaction.Configuration, diff jsondiff.DiffSet, fieldOwner string) (*ssa.ChangeSet, error) {
+	cfg, err := config.RESTClientGetter.ToRESTConfig()
+	if err != nil {
+		return nil, err
+	}
+	c, err := client.New(cfg, client.Options{})
+	if err != nil {
+		return nil, err
+	}
+
+	var toCreate, toPatch sortableDiffs
+	for _, d := range diff {
+		switch d.Type {
+		case jsondiff.DiffTypeCreate:
+			toCreate = append(toCreate, d)
+		case jsondiff.DiffTypeUpdate:
+			toPatch = append(toPatch, d)
+		}
+	}
+
+	var (
+		changeSet = ssa.NewChangeSet()
+		errs      []error
+	)
+
+	sort.Sort(toCreate)
+	for _, d := range toCreate {
+		obj := d.DesiredObject.DeepCopyObject().(client.Object)
+		if err := c.Create(ctx, obj, client.FieldOwner(fieldOwner)); err != nil {
+			errs = append(errs, fmt.Errorf("%s creation failure: %w", subjectFromObject(obj), err))
+			continue
+		}
+		changeSet.Add(objectToChangeSetEntry(obj, ssa.CreatedAction))
+	}
+
+	sort.Sort(toPatch)
+	for _, d := range toPatch {
+		data, err := json.Marshal(d.Patch)
+		if err != nil {
+			errs = append(errs, fmt.Errorf("%s patch failure: %w", subjectFromObject(d.DesiredObject), err))
+			continue
+		}
+
+		obj := d.DesiredObject.DeepCopyObject().(client.Object)
+		patch := client.RawPatch(types.JSONPatchType, data)
+		if err := c.Patch(ctx, obj, patch, client.FieldOwner(fieldOwner)); err != nil {
+			if obj.GetObjectKind().GroupVersionKind().Kind == "Secret" {
+				err = maskSensitiveErrData(err)
+			}
+			errs = append(errs, fmt.Errorf("%s patch failure: %w", subjectFromObject(obj), err))
+			continue
+		}
+		changeSet.Add(objectToChangeSetEntry(obj, ssa.ConfiguredAction))
+	}
+
+	return changeSet, apierrutil.NewAggregate(errs)
+}
+
+// objectToChangeSetEntry returns a ssa.ChangeSetEntry for the given object and
+// action.
+func objectToChangeSetEntry(obj client.Object, action ssa.Action) ssa.ChangeSetEntry {
+	return ssa.ChangeSetEntry{
+		ObjMetadata: object.ObjMetadata{
+			GroupKind: obj.GetObjectKind().GroupVersionKind().GroupKind(),
+			Name:      obj.GetName(),
+			Namespace: obj.GetNamespace(),
+		},
+		GroupVersion: obj.GetObjectKind().GroupVersionKind().Version,
+		Subject:      subjectFromObject(obj),
+		Action:       action,
+	}
+}
+
+// subjectFromObject returns a string representation of the given object.
+func subjectFromObject(obj client.Object) string {
+	return fmt.Sprintf("%s/%s/%s", obj.GetObjectKind().GroupVersionKind().Kind, obj.GetNamespace(), obj.GetName())
```

Comment:
I would use the SSA formatter that knows how to deal with non namespaced objects.

Reply:
Can not use that here because I do not have a `object.ObjMetadata` at all times, but apply the same logic. Thanks :bow: